File: app/src/main/java/com/haishinkit/rtmp/RTMPSocket.java
package com.haishinkit.rtmp;

import java.nio.ByteBuffer;
import com.haishinkit.net.Socket;
import com.haishinkit.util.Log;
import com.haishinkit.rtmp.messages.RTMPMessage;

import org.apache.commons.lang3.builder.ToStringBuilder;

public final class RTMPSocket extends Socket {
    enum ReadyState {
        Uninitialized,
        VersionSent,
        AckSent,
        HandshakeDone,
        Closing,
        Closed
    }

    private int bandwidth = 0;
    private int chunkSizeC = RTMPChunk.DEFAULT_SIZE;
    private int chunkSizeS = RTMPChunk.DEFAULT_SIZE;
    private boolean connected = false;
    private RTMPHandshake handshake = new RTMPHandshake();
    private ReadyState readyState = ReadyState.Uninitialized;
    private RTMPConnection connection = null;

    public RTMPSocket(final RTMPConnection connection) {
        this.connection = connection;
    }

    public boolean isConnected() {
        return connected;
    }

    public int getBandWidth() {
        return bandwidth;
    }

    public RTMPSocket setBandwidth(final int bandwidth) {
        this.bandwidth = bandwidth;
        return this;
    }

    public int getChunkSizeS() {
        return chunkSizeS;
    }

    public RTMPSocket setChunkSizeS(final int chunkSizeS) {
        this.chunkSizeS = chunkSizeS;
        return this;
    }

    public int getChunkSizeC() {
        return chunkSizeC;
    }

    public RTMPSocket setChunkSizeC(final int chunkSizeC) {
        this.chunkSizeC = chunkSizeC;
        return this;
    }

    public void doOutput(final RTMPChunk chunk, final RTMPMessage message) {
        if (chunk == null || message == null) {
            throw new IllegalArgumentException();
        }
        for (ByteBuffer buffer : chunk.encode(this, message)) {
            doOutput(buffer);
        }
    }

    @Override
    protected void onConnect() {
        Log.v(getClass().getName() + "#onConnect", "");
        chunkSizeC = RTMPChunk.DEFAULT_SIZE;
        chunkSizeS = RTMPChunk.DEFAULT_SIZE;
        handshake.clear();
        readyState = ReadyState.VersionSent;
        doOutput(handshake.getC0C1Packet());
    }

    @Override
    protected void listen(final ByteBuffer buffer) {
        switch (readyState) {
            case VersionSent:
                if (buffer.limit() < RTMPHandshake.SIGNAL_SIZE + 1) {
                    break;
                }
                handshake.setS0S1Packet(buffer);
                doOutput(handshake.getC2Packet());
                buffer.position(RTMPHandshake.SIGNAL_SIZE + 1);
                readyState = ReadyState.AckSent;
                if (buffer.limit() - buffer.position() == RTMPHandshake.SIGNAL_SIZE) {
                    listen(buffer.slice());
                    buffer.position(3073);
                }
                break;
            case AckSent:
                if (buffer.limit() < RTMPHandshake.SIGNAL_SIZE) {
                    break;
                }
                handshake.setS2Packet(buffer);
                buffer.position(RTMPHandshake.SIGNAL_SIZE);
                readyState = ReadyState.HandshakeDone;
                connected = true;
                doOutput(RTMPChunk.ZERO, connection.createConnectionMessage());
                break;
            case HandshakeDone:
                try {
                    connection.listen(buffer);
                } catch (IndexOutOfBoundsException e) {

                } catch (IllegalArgumentException e) {
                    e.printStackTrace();
                    throw e;
                }
                break;
            default:
                break;
        }
    }

    public String toString() {
        return ToStringBuilder.reflectionToString(this);
    }
}

File: haishinkit/src/main/java/com/haishinkit/rtmp/RtmpSocket.kt
package com.haishinkit.rtmp

import android.util.Log
import com.haishinkit.event.Event
import com.haishinkit.net.NetSocket
import com.haishinkit.net.NetSocketImpl
import java.nio.BufferUnderflowException
import java.nio.ByteBuffer

internal class RtmpSocket(val connection: RtmpConnection) : NetSocket.Listener {
    enum class ReadyState {
        Uninitialized,
        VersionSent,
        AckSent,
        HandshakeDone,
        Closing,
        Closed
    }

    var bandWidth = 0
        internal set
    var chunkSizeC = RtmpChunk.DEFAULT_SIZE
    var chunkSizeS = RtmpChunk.DEFAULT_SIZE
    var isConnected = false
        private set
    var timeout: Int
        get() = socket?.timeout ?: 0
        set(value) {
            socket?.timeout = value
        }
    val totalBytesIn: Long
        get() = socket?.totalBytesIn?.get() ?: 0
    val totalBytesOut: Long
        get() = socket?.totalBytesOut?.get() ?: 0
    private var socket: NetSocket? = null
    private val handshake: RtmpHandshake by lazy {
        RtmpHandshake()
    }
    private var readyState = ReadyState.Uninitialized
        set(value) {
            field = value
            connection.onSocketReadyStateChange(this, value)
        }

    fun connect(dstName: String, dstPort: Int, isSecure: Boolean) {
        socket?.listener = null
        socket = NetSocketImpl()
        socket?.listener = this
        socket?.connect(dstName, dstPort, isSecure)
    }

    fun doOutput(buffer: ByteBuffer) {
        socket?.doOutput(buffer)
    }

    fun createByteBuffer(capacity: Int): ByteBuffer {
        return socket?.createByteBuffer(capacity) ?: ByteBuffer.allocate(capacity)
    }

    fun close(disconnected: Boolean) {
        if (!isConnected) return
        var data: Any? = null
        if (disconnected) {
            data = if (readyState == ReadyState.HandshakeDone) {
                RtmpConnection.Code.CONNECT_CLOSED.data("")
            } else {
                RtmpConnection.Code.CONNECT_FAILED.data("")
            }
        }
        readyState = ReadyState.Closing
        socket?.close(disconnected)
        data?.let {
            connection.dispatchEventWith(Event.RTMP_STATUS, false, it)
        }
        readyState = ReadyState.Closed
        isConnected = false
    }

    override fun onTimeout() {
        close(false)
        connection.dispatchEventWith(Event.IO_ERROR, false)
        Log.i(TAG, "a connection was timeout")
    }

    override fun onConnect() {
        chunkSizeC = RtmpChunk.DEFAULT_SIZE
        chunkSizeS = RtmpChunk.DEFAULT_SIZE
        handshake.clear()
        readyState = ReadyState.VersionSent
        socket?.doOutput(handshake.c0C1Packet)
    }

    override fun onClose(disconnected: Boolean) {
        close(disconnected)
    }

    override fun onInput(buffer: ByteBuffer) {
        when (readyState) {
            ReadyState.VersionSent -> {
                if (buffer.limit() < RtmpHandshake.SIGNAL_SIZE + 1) {
                    return
                }
                handshake.s0S1Packet = buffer
                socket?.doOutput(handshake.c2Packet)
                buffer.position(RtmpHandshake.SIGNAL_SIZE + 1)
                readyState = ReadyState.AckSent
                if (buffer.limit() - buffer.position() == RtmpHandshake.SIGNAL_SIZE) {
                    onInput(buffer.slice())
                    buffer.position(3073)
                }
            }
            ReadyState.AckSent -> {
                if (buffer.limit() < RtmpHandshake.SIGNAL_SIZE) {
                    return
                }
                handshake.s2Packet = buffer
                buffer.position(RtmpHandshake.SIGNAL_SIZE)
                readyState = ReadyState.HandshakeDone
                isConnected = true
                connection.uri?.let {
                    connection.doOutput(RtmpChunk.ZERO, connection.createConnectionMessage(it))
                }
            }
            ReadyState.HandshakeDone ->
                try {
                    connection.listen(buffer)
                } catch (e: IndexOutOfBoundsException) {
                    if (VERBOSE) Log.d(TAG, "", e)
                } catch (e: BufferUnderflowException) {
                    if (VERBOSE) Log.d(TAG, "", e)
                } catch (e: IllegalArgumentException) {
                    if (VERBOSE) Log.w(TAG, "", e)
                    throw e
                }
            else -> {}
        }
    }

    companion object {
        private const val VERBOSE = false
        private var TAG = RtmpSocket::class.java.simpleName
    }
}
